anyblok_wms_base/inventory/node.py

```diff
@@ -0,0 +1,380 @@
+                              .all())
+        if non_ready_children:
+            raise NodeChildrenStateError(
+                self, non_ready_children,
+                "This {node} has some non children "
+                "who didn't push their Actions to it yet: "
+                "{children_states!r}")
+
+        PhysObj = self.registry.Wms.PhysObj
+        POType = PhysObj.Type
+        Avatar = PhysObj.Avatar
+        Inventory = self.registry.Wms.Inventory
+        Line = Inventory.Line
+        Action = Inventory.Action
+
+        excluded_types = self.inventory.excluded_types
+        if not excluded_types:
+            phobj_filter = None
+        else:
+            def phobj_filter(query):
+                excluded_types_q = (POType.query(POType.id)
+                                    .filter(POType.code.in_(excluded_types)))
+                return query.filter(not_(PhysObj.type_id.in_(excluded_types_q)))
+
+        cols = (Avatar.location_id, PhysObj.code, PhysObj.type_id)
+        quantity_query = self.registry.Wms.quantity_query
+        existing_phobjs = (quantity_query(location=self.location,
+                                          location_recurse=self.is_leaf,
+                                          additional_filter=phobj_filter)
+                           .add_columns(*cols).group_by(*cols)
+                           .subquery())
+
+        node_lines = (Line.query(Line.quantity,
+                                 Line.location_id,
+                                 Line.type_id, Line.code, Line.properties)
+                      .filter_by(node=self).subquery())
+        comp_query = (
+            self.registry.query(node_lines)
+            .join(existing_phobjs,
+                  # multiple criteria to join on the subquery would fail,
+                  # complaining of lack of foreign key (SQLA bug maybe)?
+                  # but it works with and_()
+                  and_(node_lines.c.type_id == existing_phobjs.c.type_id,
+                       node_lines.c.location_id == (existing_phobjs.c
+                                                    .location_id),
+                       or_(node_lines.c.code == existing_phobjs.c.code,
+                           and_(existing_phobjs.c.code.is_(None),
+                                node_lines.c.code.is_(None)))),
+                  full=True)
+            .filter(func.coalesce(existing_phobjs.c.qty, 0) !=
+                    func.coalesce(node_lines.c.quantity, 0))
+            .add_columns(func.coalesce(existing_phobjs.c.qty, 0)
+                         .label('phobj_qty'),
+                         # these columns are useful only if lines is None:
+                         existing_phobjs.c.location_id.label('phobj_loc'),
+                         existing_phobjs.c.type_id.label('phobj_type'),
+                         existing_phobjs.c.code.label('phobj_code'),
+                         ))
+
+        for row in comp_query.all():
```

Comment:
@gracinet 
This is inserting Inventory.Action for POType that have also the 'container' behaviour.
Should we filter the query to don't ask for the POType with the 'container' behaviour ?
For the moment i used the Inventory.excluded_types field to make it work as i expected.

Reply:
In a nutshell, yes that was my intent with this `excluded_types` field. Excluding all containers would be too wide in my opinion: there are use cases where containers also have to be inventoried, and I don't want AWB to assume otherwise.

Reply:
Ok I understand and I agree, i will keep using `excluded_types` field then.